server/commands/check.go

```diff
@@ -574,26 +585,29 @@ func (q *CheckQuery) resolveTupleToUserset(
 			}
 
 			select {
-			case c <- &chanResolveResult{err: err, found: nestedRC.userFound()}:
+			case c <- err:
 			case <-done:
 			}
-		}(c)
+		}(errorChannel)
 	}
 
 	// If any `break` was triggered, immediately release any possible resources held by the iterator.
 	iter.Stop()
 
-	go func(c chan *chanResolveResult) {
+	go func(errorChannel chan error) {
 		wg.Wait()
-		close(c)
-	}(c)
-
-	for res := range c {
-		if res.found {
+		close(errorChannel)
+	}(errorChannel)
+
+	// if the user is found we swallow any errors that may have be thrown
+	// otherwise we return the last error we found (if any)
+	for potentialError := range errorChannel {
+		if rc.userFound() {
+			go utils.DrainChannel(errorChannel)
```

Comment:
Why drain the channel instead of just returning? Is there a way to make this code more easily understandable? It's not obvious why we're calling `DrainChannel` on the same error channel that we're already iterating over, and it seems quite likely that the `DrainChannel` could hang indefinitely waiting on the read of the error channel.

Reply:
It's the forgotten sender problem... When we return here, we stop consuming `errorChannel`, which means that any goroutine that tries to write to the channel will become blocked. This is because `errorChannel` is unbuffered.

> Is there a way to make this code more easily understandable?

We don't know :( If there is, feel free to send an improvement